### PR TITLE
fix: make atto parser faster than the scala parser combinator one

### PR DESCRIPTION
**Pull request by mbaechler**
For some reason, I'd rather replace the scala parser combinator implementation with something else.

I read that the atto parser is not the default one because it's slower than the current one ( https://github.com/alonsodomin/cron4s/issues/468#issuecomment-1617526076 )

This PR fixes some correctness issues and makes it faster than the current one, it may be enough to allow the switch?

Here are my numbers regarding performances

```
attoParser before                                10-35 2,4,6 * ? * *  avgt   20  69.297 ± 0.223  us/op
attoParser after                                 10-35 2,4,6 * ? * *  avgt   10  43.819 ± 2.469  us/op
parserCombinators                                10-35 2,4,6 * ? * *  avgt   20  43.239 ± 0.148  us/op

attoParser before  * 5,10,15,20,25,30,35,40,45,50,55/2 * ? * mon-fri  avgt   20  82.511 ± 0.298  us/op
attoParser after   * 5,10,15,20,25,30,35,40,45,50,55/2 * ? * mon-fri  avgt   10  45.966 ± 0.409  us/op
parserCombinators  * 5,10,15,20,25,30,35,40,45,50,55/2 * ? * mon-fri  avgt   20  60.507 ± 0.212  us/op

attoParser before                                    10-65 * * * * *  avgt   20  13.337 ± 0.052  us/op
attoParser after                                     10-65 * * * * *  avgt   10   9.366 ± 0.053  us/op
parserCombinators                                    10-65 * * * * *  avgt   20  26.571 ± 0.117  us/op

attoParser before                            * */10 5-10 ? * mon-fri  avgt   20  62.982 ± 0.262  us/op
attoParser after                             * */10 5-10 ? * mon-fri  avgt   10  32.665 ± 0.265  us/op
parserCombinators                            * */10 5-10 ? * mon-fri  avgt   20  42.597 ± 0.214  us/op

attoParser before     */30 10,20,40 5-15,25-35/4 ? 1,3,7,oct-dec sun  avgt   20  38.053 ± 0.194  us/op
attoParser after      */30 10,20,40 5-15,25-35/4 ? 1,3,7,oct-dec sun  avgt   10  26.602 ± 0.175  us/op
parserCombinators     */30 10,20,40 5-15,25-35/4 ? 1,3,7,oct-dec sun  avgt   20  47.154 ± 0.181  us/op
```

### CODE REVIEW

File: bench/src/main/scala/cron4s/atto/package.scala [hunk 1]
```diff
@@ -19,16 +19,33 @@ package cron4s
...
 import cron4s.expr._
 
 package object atto {
   import CronField._
   import CronUnit._
 
-  private val sexagesimal: Parser[Int] = int.filter(x => x >= 0 && x < 60)
-  private val decimal: Parser[Int]     = int.filter(x => x >= 0)
-  private val literal: Parser[String]  = takeWhile1(_ >= ' ')
+  private def oneOrTwoDigitsPositiveInt: Parser[Int] = {
```
Comment (mbaechler) [inline bench/src/main/scala/cron4s/atto/package.scala:28]: this is the optimization that makes it much faster

File: bench/src/main/scala/cron4s/atto/package.scala [hunk 2]
```diff
@@ -19,16 +19,33 @@ package cron4s
...
+        // scala-js can't parse non-alpha digits so we just fail in that case.
+        case _: java.lang.NumberFormatException =>
+          err[Int]("https://github.com/scala-js/scala-js/issues/2935")
+      }
+    )
+  } namedOpaque "oneOrTwoDigitsPositiveInt"
+
+  private val sexagesimal: Parser[Int] = oneOrTwoDigitsPositiveInt.filter(x => x >= 0 && x < 60)
+
+  private val literal: Parser[String]  = takeWhile1(x => x != ' ' && x != '-')
```
Comment (mbaechler) [inline bench/src/main/scala/cron4s/atto/package.scala:48]: This is a correctness fix as things like `mon-fri` was failing previously

File: bench/src/main/scala/cron4s/atto/package.scala [hunk 3]
```diff
@@ -111,9 +128,10 @@ package object atto {
       unit: CronUnit[F]
   ): Parser[SeveralNode[F]] = {
     def compose(b: => Parser[EnumerableNode[F]]) =
-      sepBy1(b, comma)
-        .filter(_.size > 1)
-        .map(values => SeveralNode.fromSeq[F](values.toList).get)
+      sepBy(b, comma)
```
Comment (mbaechler) [inline bench/src/main/scala/cron4s/atto/package.scala:131]: We were actually checking the size several times, now it's done a single time

File: bench/src/main/scala/cron4s/atto/package.scala [hunk 4]
```diff
@@ -166,8 +184,9 @@ package object atto {
   } yield CronExpr(sec, min, hour, day, month, weekDay)
 
   def parse(e: String): Either[Error, CronExpr] =
-    (cron.parseOnly(e): @unchecked) match {
+    (phrase(cron).parseOnly(e): @unchecked) match {
       case ParseResult.Done(_, result) => Right(result)
+      case ParseResult.Fail("", _, _) => Left(ExprTooShort)
```
Comment (mbaechler) [inline bench/src/main/scala/cron4s/atto/package.scala:189]: This is a correctness fix : a too short expression was not detected as such

File: bench/src/main/scala/cron4s/atto/package.scala [hunk 5]
```diff
@@ -166,8 +184,9 @@ package object atto {
   } yield CronExpr(sec, min, hour, day, month, weekDay)
 
   def parse(e: String): Either[Error, CronExpr] =
-    (cron.parseOnly(e): @unchecked) match {
+    (phrase(cron).parseOnly(e): @unchecked) match {
```
Comment (mbaechler) [inline bench/src/main/scala/cron4s/atto/package.scala:187]: this is a correctness fix : it ensure the string is parsed entirely

**Comment by mbaechler** [comment]
Hi @alonsodomin , it seems you are not active anymore on github and cron4s, any chance you consider discussing about sharing the maintenance of cron4s project?

**Comment by alonsodomin** [comment]
hello @mbaechler, thanks for your submitted work and sorry for reviewing this earlier (btw, best way to nudge me into reviewing something is to add me as a reviewer in the PR). Anyway, I'm on and off at Github but definitely more active on other repos than this one.

I have no issue on sharing ownership of this repo, it's indeed useful to the community (even though I don't use it as much) and I think it would be great to do a clean up of the library. I'll email you my contact so we can continue the conversation.

Regarding the changes addressed in this PR, how confident you feel after this improvements to make Atto the default parser? In my view this should move in that direction but also there should be a period of time in which the scala combinators parser is still available and that users could choose to use it (before finally deleting it from the code).

**Comment by mbaechler** [comment]
Thank you for your answer.

> Regarding the changes addressed in this PR, how confident you feel after this improvements to make Atto the default parser?

I'm not very confident, right now.
I wanted to see if lifting some barriers would make that hypothesis acceptable.
If it is, I would first implement a property test that compare parsing results for both parser on a wide variety of inputs (think golden master testing)
Once proven that both are doing the same thing, I would propose a switch.

> In my view this should move in that direction but also there should be a period of time in which the scala combinators parser is still available and that users could choose to use it (before finally deleting it from the code).

It sounds reasonable.

**Comment by alonsodomin** [comment]
in that case, it may be worth moving the atto impl out of the bench code and into a separate module and make the test kit target it to at least ensure that functionality between the two is fully preserved.

**Comment by mbaechler** [comment]
> in that case, it may be worth moving the atto impl out of the bench code and into a separate module and make the test kit target it to at least ensure that functionality between the two is fully preserved.

I will do in another commit, do you think we can merge this one?

**Comment by mbaechler** [comment]
> > in that case, it may be worth moving the atto impl out of the bench code and into a separate module and make the test kit target it to at least ensure that functionality between the two is fully preserved.
> 
> I will do in another commit, do you think we can merge this one?

I guess, as it touches a single file that is not a part of any library artifact, I can merge it safely.

**Pull request merged**